src/libprojectM/PCM.cpp

```diff
@@ -333,20 +333,40 @@ void PCM::getPCM(float *PCMdata, int samples, int channel, int freq, float smoot
        PCMdata[samples-1]=0;
      }
 
-   //return frequency data instead of PCM (perform FFT)
-
-   if (freq)
-     {
-       // NOTE some presets set bSpectrum=1 and samples!=2^n, not sure what rdft() does with that
-       assert(samples <= 1024);
-       samples = std::min(1024,samples);
-       double temppcm[1024];
-       for (int i=0;i<samples;i++)
-         {temppcm[i]=(double)PCMdata[i];}
-       rdft(samples, 1, temppcm, ip, w);
-       for (int j=0;j<samples;j++)
-         {PCMdata[j]=(float)temppcm[j];}
-     }
+		 // return frequency data instead of PCM (perform FFT)
+
+		 if (freq)
+		 {
+			 // NOTE some presets set bSpectrum=1 and samples!=2^n, not sure what rdft() does with that
+
+			 // samples needs to be a power of 2 or rdft will blow up
+			 int samples_pow2 = samples;
+			 if ((samples & (samples - 1)) != 0)
+			 {
+				 int power = 1;
+				 // round down to nearest power of 2
+				 while (samples_pow2 >>= 1)
+					 power <<= 1;
+				 samples_pow2 = power;
+			 }
+
+			 double temppcm[1024];
+			 samples_pow2 = std::min(samples_pow2, 1024);
+			 for (int i = 0; i < samples; i++)
+			 {
+				 // fft input
+				 temppcm[i] = (double)PCMdata[i];
+			 }
+			 // fft
+			 rdft(samples_pow2, 1, temppcm, ip, w);
+			 for (int j = 0; j < samples; j++)
+			 {
+				 // get fft output, if we had to get a fewer number of samples than requested
+				 // then we just fill in the rest with 0. this is lame.
+				 // TODO: better solution if requested number of samples is not a power of 2
```

Comment:
I feel like we're sacrificing some accuracy here

Reply:
I see.  This happens when a waveform has bspectrum==true and samples set to a strange size.  You know this class is constantly recomputing the spectrum data on every call to addPCM().  It's sitting in vdataL and vdataR.  

Maybe it makes more sense to change CustomWave read directly from pcmdataL/R or vdataL?R?

Reply:
That's not quite right, because Wafeform uses the smoothing parameter.  (PCM has an awkward interface).  In any event, maybe it makes sense to use vdataL when spectrum==true?

Doing an fft on 32 samples doesn't make much sense...  Grabbing the first 32 values out of the larger FFT seems more useful.

Reply:
I think we can postpone the PCM.cpp part of this PR, and maybe look into changing how Waveform to pulls data.  

I'm reviewing how code uses PCM and will look at that.  BTW, I noticed that BeatDetect is _still_ pulling from pcmdata instead of vdata.  Didn't catch that with my previous beatdetect change... Look for PR maybe next weekend?

Reply:
Gotcha ok. Feel free to edit/amend this PR. Come chat on Discord if you want!

Reply:
Should I undo these changes?

Reply:
Yes, undo the PCM change for now.  I'll look at the whole Waveform/PCM behavior as one piece and with an eye toward milkdrop2 behavior.

Reply:
Ok done. That would be great. I think right now this is a pretty urgent security/stability issue